### PR TITLE
king: commit seppuku when server is out of fds, also setrlimit allowing for more fds

### PR DESCRIPTION
**Pull request by pilfer-pandex**
Following joe's suggested response to #4530, we shut down the king when the http server is saturated with requests, as an alternative to getting the server permanently stuck in a state where it cannot accept new connections.

To make this condition far less likely, we increase the resource cap on file descriptors from 256 to 10240 (the value of OPEN_MAX, at least on darwin).

In the long run, we should probably try to restart the http server rather than shut down the king, or selectively go through and terminate specific requests.

### CODE REVIEW

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs [hunk 1]
```diff
@@ -356,7 +361,11 @@ eyre env who plan isFake stderr sub = (initialEvents, runHttpServer)
   restart :: Drv -> HttpServerConf -> RIO e Serv
   restart (Drv var) conf = do
     logInfo "Restarting http server"
-    let startAct = startServ who isFake conf plan stderr sub
+    let onFatal = runRIO env $ do
+          -- XX instead maybe restart following logic under HSESetConfig below
+          stderr "You've been DDoSed. Please restart your ship."
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs:366]: [citation needed]

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs [hunk 2]
```diff
@@ -254,19 +255,28 @@ startServer
...
   let host = case hos of
         SHLocalhost -> "127.0.0.1"
         SHAnyHostOk -> "*"
 
+  let handler r e
+        | Just (IOError {ioe_type = ResourceExhausted}) <- fromException e
+        = runRIO envir $ do
+            logError (displayShow e)
+            io onFatal
+        | otherwise = W.defaultOnException r e
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:272]: What does the default exception do? Is that what's currently printing and then hanging?

**Review by joemfb — CHANGES_REQUESTED** [review]
I think we should print the actual error before shutting down, and I'm not sure we want to be deferring to the default exception handler.

Also, there's a build issue with this PR. Look like maybe you missed an Eyre call site?

**Comment by joemfb** [comment]
As for `OPEN_MAX` it's a constant `10240`, and has been for quite awhile. FFI seems like serious overkill; it'd probably be fine to just hardcode it with a comment.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs (hunk 1 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs:366]: [S. T. Zargar, J. Joshi and D. Tipper, "A Survey of Defense Mechanisms Against Distributed Denial of Service (DDoS) Flooding Attacks," in IEEE Communications Surveys & Tutorials, vol. 15, no. 4, pp. 2046-2069, Fourth Quarter 2013, doi: 10.1109/SURV.2013.031413.00127.](https://sci-hub.se/https://ieeexplore.ieee.org/abstract/document/6489876)

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs (hunk 2 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:272]: The default handler is found at https://hackage.haskell.org/package/warp-3.3.14/docs/src/Network.Wai.Handler.Warp.Settings.html#defaultOnException. It either prints or does not print the exception depending on the type, then discards it. Accordingly, it's the origin of the printout @lycurgus reported. It's not the source of the hang; that's because warp gets into a bad state when it runs out of file descriptors; it's not a generic thing that happens on any exception we might encounter here, nor does "the exception handler hang" as you describe it.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs [hunk 3]
```diff
@@ -254,19 +255,28 @@ startServer
...
   envir <- ask
 
   let host = case hos of
         SHLocalhost -> "127.0.0.1"
         SHAnyHostOk -> "*"
 
+  let handler r e
+        | Just (IOError {ioe_type = ResourceExhausted}) <- fromException e
+        = runRIO envir $ do
+            logError (displayShow e)
```
Comment (pilfer-pandex) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:270]: Replying to your top-level comment: We do print the error before shutting down, and do so right here. Do you want this to be to stderr in addition to (or instead of) the logError?

**Review by pilfer-pandex — COMMENTED** [review]
> it's a constant 10240

Strange. I got what I think was EINVAL when I tried even 1023. (I was trying to do binary search to discover it lol.)

Let me try again.

**Comment by pilfer-pandex** [comment]
Holy shit my Stack cache was dirty. That's why the build error got through. It's holding onto previous copies of compiled modules as though they were the current ones.

Also your suggestion of 10240 worked. I suspect it was just too late last night and I forgot to copy the executable when I did my test.

**Comment by pilfer-pandex** [comment]
To confirm, do we also want 10240 on linux?

**Comment by pilfer-pandex** [comment]
@joemfb, build problem this time seems to be in nix itself; there's a printout inviting me to report a bug to nixos, to wit:

```

installing 'nss-cacert-3.49.2'
error: unexpected EOF reading a line

---- oh no! --------------------------------------------------------------------
Jeeze, something went wrong. If you can take all the output and open
an issue, we'd love to fix the problem so nobody else has this issue.
```

Accordingly I think you can proceed with the review. @brendanhay have you seen this problem before?

**Comment by pilfer-pandex** [comment]
Seems to be a nondeterministic failure that you just get occasionally if you use macos?? https://github.com/NixOS/nix/issues/3605. "Jeeze" indeed.

File: pkg/hs/urbit-king/lib/Urbit/King/Main.hs [hunk 4]
```diff
@@ -693,6 +703,11 @@ main = do
     for_ [Sys.sigTERM, Sys.sigINT] $ \sig -> do
       Sys.installHandler sig (Sys.Catch onKillSig) Nothing
 
+  setRLimits = do
+    nofiles <- Sys.getResourceLimit Sys.ResourceOpenFiles
+    Sys.setResourceLimit Sys.ResourceOpenFiles
+      nofiles { Sys.softLimit = Sys.ResourceLimit 10240 }
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/King/Main.hs:709]: We still want the max number of files on linux, which should be accurately reported by `getResourceLimit` and can be higher than this. It just needs to be hardcoded on MacOS.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs (hunk 1 above)
Comment (joemfb) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre.hs:366]: I don't think this error message is appropriate. If this is to be a generic error handler, what it print should be generic. If it's specifically meant to handle the ResourceLimit exception caused by leaking sockets, it shouldn't make grandiose claims about cyberwarfare.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs (hunk 2 above)
Comment (joemfb) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:272]: Do you have any indication that printing/ignoring exceptions is the right default behavior here? It seems like it'd be better to whitelist non-fatal exceptions and otherwise `abort()`.

**Review by joemfb — CHANGES_REQUESTED** [review]

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs (hunk 2 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:272]: Yeah, I do. I looked for examples of exception handling behavior in major open source codebases using warp. Take a look at:

- https://github.com/hasura/graphql-engine/blob/master/server/src-lib/Hasura/App.hs#L583
- https://github.com/PostgREST/postgrest/blob/main/main/Main.hs#L89

Both of these major, legitimate applications *fail to modify the default exception handler at all*, so they definitely don't `abort()` on any exception. 

Now the comment at https://hackage.haskell.org/package/warp-3.3.14/docs/src/Network.Wai.Handler.Warp.Settings.html#defaultShouldDisplayException says

> The goal is to hide exceptions which occur under the normal course of the web server running.

So perhaps the list found in that function will serve as a suitable whitelist? But in fact, the stderr printouts reported in https://github.com/urbit/urbit/issues/4613 are (I've concluded) coming out of this exception handler, and I think we can agree that these should not shut down the server.

Because of the (unfortunate) nature of the Haskell exception mechanism, it's not possible to confidently get a stable list of all the exceptions that might trip this handler, let alone fix a subset of those such that we can be sure its complement contains only things worth shutting down the server over. I'm just not comfortable implementing a policy like you propose and putting it on zod. I'd have to read all of the upstream libraries to find all places where exceptions are thrown, and I'd have to be confident that that doesn't change over time.

I'm going to issue a change here which uses the logger instead of stderr for printing exceptions, and maybe adds more info. (That's a good fix for the linked issue.)

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs [hunk 5]
```diff
@@ -254,19 +255,38 @@ startServer
...
+        Just r  -> "eyre: failed request from " <> (tshow $ W.remoteHost r)
+                <> " for " <> (tshow $ W.rawPathInfo r) <> ": " <> tshow e
+        Nothing -> "eyre: server exception: " <> tshow e
+
   let opts =
         W.defaultSettings
           & W.setHost host
           & W.setPort (fromIntegral por)
-          & W.setTimeout (5 * 60)
+          & W.setTimeout 30
```
Comment (joemfb) [inline pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:288]: This timeout might be a little too aggressive, but it'll be easy to tune it later.

**Review by joemfb — APPROVED** [review]
LGTM

File: pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs (hunk 5 above)
Comment (pilfer-pandex) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Eyre/Serv.hs:288]: IIRC it's the timeout vere uses.
